### PR TITLE
ENH: improve dtype check in wavfile.write

### PR DESCRIPTION
**Pull request by JuanFMontesinos**
#### What does this implement/fix?
`scipy.io.wavfile.write` has a wrong data type check. It checks a numpy array to be "float" or "integer" in a generic way. However float16 is not supported by any codec but accepted by the function, leading to a silent error.
Likewise, exotic data types such as float128, int128 and so on are neither supported. 
#### Additional information  
While it's complicated to find what dtypes are exactly supported, I rely on ffmpeg formats as they are a superset of scipy's ones.  
int64 is also supported despite not appearing. 
image  
From your references  
image

### CODE REVIEW

File: scipy/io/wavfile.py [hunk 1]
```diff
@@ -770,8 +770,9 @@ def write(filename, rate, data):
 
     try:
         dkind = data.dtype.kind
-        if not (dkind == 'i' or dkind == 'f' or (dkind == 'u' and
-                                                 data.dtype.itemsize == 1)):
+        allowed_dtypes = ['float32', 'float64',
+                          'uint8', 'int8', 'int16', 'int32', 'int64']
```
[the remark this answers is not in the document]
Comment (JuanFMontesinos) [reply scipy/io/wavfile.py:774]: Well, I don't know the intrinsecs but the docs only show common data types, not all data types. 
The second table is extracted from the references provided in the docs:
IBM Corporation and Microsoft Corporation, “Multimedia Programming Interface and Data Specifications 1.0”, section “Data Format of the Samples”, August 1991 http://www.tactilemedia.com/info/MCI_Control_Info.html

I need to recheck but I think I did a quick test and ffprobe was accepting the output for int8.

[the remark this answers is not in the document]
Comment (JuanFMontesinos) [reply scipy/io/wavfile.py:774]: image
You are right, according to ffprobe info it seems it's processed as a uint8.

**Comment by j-bowhay** [comment]
@JuanFMontesinos could you add the suggested regression test, please?

**Comment by JuanFMontesinos** [comment]
@j-bowhay 

Done at 37f0d17372060ad2ea96700cd164a902f257197b! Although tests are failing due to an import error out of my control.

**Comment by lucascolley** [comment]
Hey @JuanFMontesinos , would you mind re-triggering CI here? Pushing an empty commit will do it.

**Comment by JuanFMontesinos** [comment]
> Hey @JuanFMontesinos , would you mind re-triggering CI here? Pushing an empty commit will do it.

Done!

**Comment by lucascolley** [comment]
Okay, the docs build should pass if you merge main into this branch. Then a maintainer can approve the GHA workflows.

**Review by lucascolley — APPROVED** [review]
CI is happy, the change is simple enough and the regression test is there, so pulling this in. Thanks for your first contribution to SciPy @JuanFMontesinos ! And thanks for the test @tylerjereddy

**Comment by JuanFMontesinos** [comment]
@lucascolley @tylerjereddy  Thank you very much and thanks all the developers for such a nice work with SciPy.

**Pull request merged**